rfc/rfc-0002-bindings.md

```diff
@@ -0,0 +1,97 @@
+
+riff should adopt CNB Bindings spec for bindings at both build and runtime.
+
+Kubernetes uses [Volumes](https://kubernetes.io/docs/concepts/storage/volumes/) to inject files from an external source into a running container. While there is a diverse spectrum of volume implementations available, in order to maximize both compatibility with bog standard Kubernetes and simplify the implementation, we should focus on a [configMap volume](https://kubernetes.io/docs/concepts/storage/volumes/#configmap) for binding metadata and a [secret volume](https://kubernetes.io/docs/concepts/storage/volumes/#secret) for binding secrets, which map directly to Kubernetes [ConfigMaps](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/) and [Secrets](https://kubernetes.io/docs/concepts/configuration/secret/) respectively.
+
+As a binding is backed by a ConfigMap and Secret a mechanism is needed to specify which ConfigMap and Secret. [ObjectReferences](https://godoc.org/k8s.io/api/core/v1#ObjectReference) are commonly used to define the coordinates to an arbitrary resources. A binding reference consists of metadata ObjectReference and a secret ObjectReference.
+
+Any resource that produces a binding should include the binding reference on it's status, for example:
+
+```yaml
+name: provider
+...
+status:
+  ...
+  binding:
+    metadataRef:
+      apiVersion: v1
+      kind: ConfigMap
+      namespace: default
+      name: provider-binding-metadata
+    secretRef:
+      apiVersion: v1
+      kind: Secret
+      namespace: default
+      name: provider-binding-secret
+```
+
+The provider may change the values within the metadata and secret as needed, but should avoid changing the keys. The CNB Binding spec defines some well known keys that should be defined in the metadata.
+
+A workload consuming the binding can use the provider's binding, For example:
+
+```yaml
+name: consumer
+...
+spec:
+  ...
+  bindings:
+  - name: my-binding
+    metadataRef:
+      apiVersion: v1
+      kind: ConfigMap
+      namespace: default
+      name: provider-binding-metadata
+    secretRef:
+      apiVersion: v1
+      kind: Secret
+      namespace: default
+      name: provider-binding-secret
+```
+
+It is common to consume multiple bindings within a single workload. Each binding must have a unique name for the workload. The binding name is the directory the metadata and secret volumes are mounted within. While the name has no impact on the binding per se, workloads may use the name to convey semantic meaning about the intent of the binding (e.g. a primary database vs a replica).
```

Comment:
> While the name has no impact on the binding per se

Does it not? It seems indeed that the names space is flat *for all bindings* (even of different "types", ie databases and streams would use the same set of filenames under the well known directory base).

If the name does not bear any particular meaning, it seems that there is yet another level of indirection needed to convey the "meaning" of the binding. For example, I would be interested to know what a riff processor defined as
```
riff streaming processor create foo --function-ref f --input input1:stream1 --input input2:stream1 --output output1
```
would see as mounted info.

Reply:
Please wait until comments are solicited

Reply:
I don't want the text of this RFC to get into the specifics of any particular type as that would belong to follow on RFC, but you are correct that the service binding will likely not contain enough metadata for a Process to consume Streams as bindings, and that's ok.

I wouldn't expect the Processor CRD to change initially. Since the Processor reconciler is already aware of Streams, it can continue to be aware of Streams to provide an experience higher than the raw bindings. Under the hood, the processor sidecar can switch from its current scheme to use the CNB binding format with binding published for each Stream by the Stream reconciler.

It would be possible to add some encoding scheme into the binding name, but that probably doesn't provide a good experience. Specifically for Processors, what I would initially propose is that the binding name represent the stream alias, the INPUTS, OUTPUTS and OUTPUT_CONTENT_TYPES move into the binding. INPUT_NAMES and OUTPUT_NAMES continue as envvar that conveys the order and input/output categorization.

Another step could open up the function container to accept binding. This would augment the current Processor CRD.
```